packages/gator-permissions-controller/src/decodePermission/decodePermission.ts

```diff
@@ -77,6 +77,37 @@ export const identifyPermissionByEnforcers = ({
   return matchingPermissionType;
 };
 
+/**
+ * Extracts the expiry timestamp from TimestampEnforcer caveat terms.
+ *
+ * Based on the TimestampEnforcer contract encoding:
+ * - Terms are 32 bytes total (64 hex characters without '0x')
+ * - First 16 bytes (32 hex chars): timestampAfterThreshold (uint128) - must be 0
+ * - Last 16 bytes (32 hex chars): timestampBeforeThreshold (uint128) - this is the expiry
+ *
+ * @param terms - The hex-encoded terms from a TimestampEnforcer caveat
+ * @returns The expiry timestamp in seconds, or null if no valid expiry exists
+ * @throws If the terms are not exactly 32 bytes or if the timestampAfterThreshold is non-zero
+ */
+const extractExpiryFromCaveatTerms = (terms: Hex): number | null => {
+  // Validate terms length: must be exactly 32 bytes (64 hex chars + '0x' prefix = 66 chars)
+  if (terms.length !== 66) {
+    throw new Error(
+      `Invalid TimestampEnforcer terms length: expected 66 characters (0x + 64 hex), got ${terms.length}`,
+    );
+  }
+
+  const [after, before] = splitHex(terms, [16, 16]);
+
+  if (hexToNumber(after) !== 0) {
+    throw new Error('Invalid expiry: timestampAfterThreshold must be 0');
+  }
+
+  const expiry = hexToNumber(before);
+
+  return expiry === 0 ? null : expiry;
```

Comment:
if the expiry is 0 should we consider the request invalid?

I'm not sure if it's possible to specify expiry of zero within the picker, but the permission decoding logic tends to be very strict, because part of it's responsibility is to ensure that the request is a valid request from the gator permissions system.